File: src/csharp/highflyers-protocol/Parser.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Reflection;

namespace HighFlyers.Protocol
{
    public class FrameParsedEventArgs : EventArgs
    {
        public FrameParsedEventArgs(Frame frame)
        {
            ParsedFrame = frame;
        }

        public Frame ParsedFrame { get; private set; }
    }

    public class Parser<T>
    {
        public delegate void FrameParsedHandler(object sender, FrameParsedEventArgs args);

        private readonly MethodInfo buildMethod;
        private readonly List<byte> bytes = new List<byte>();
        private bool prevSentinel;

        public Parser()
        {
            buildMethod = typeof(T).GetMethod("BuildFrame", new[] {typeof(List<byte>)});

            if (buildMethod == null)
                throw new Exception("Cannot find method GetMethod in a builder class");

            LastFrame = null;
        }

        public Frame LastFrame { get; private set; }
        public event FrameParsedHandler FrameParsed;

        protected void OnFrameParsed(FrameParsedEventArgs args)
        {
            if (FrameParsed != null)
                FrameParsed(this, args);
        }

        public void AppendBytes(byte[] data)
        {
            foreach (byte b in data)
                AppendByte(b);
        }

        private void AppendByte(byte b)
        {
            bool tmpSentinel = prevSentinel;
            if (prevSentinel)
            {
                if (b == FrameParserHelper.Sentinel || b == FrameParserHelper.EndFrame)
                    bytes.Add(b);
                else
                    throw new InvalidDataException("Unexpected token " + b);
                prevSentinel = false;
            }
            else if (b == FrameParserHelper.Sentinel)
                prevSentinel = true;
            else
                bytes.Add(b);

            if (b == FrameParserHelper.EndFrame && !tmpSentinel)
                ParseFrame();
            else if (FrameParserHelper.MaxLength == bytes.Count)
            {
                bytes.Clear();
                throw new WarningException("Too many bytes without end_frame sign. Dropping data...");
            }
        }

        private void ParseFrame()
        {
            LastFrame = buildMethod.Invoke(null, new object[] {bytes}) as Frame;
            OnFrameParsed(new FrameParsedEventArgs(LastFrame));
            bytes.Clear();
        }
    }
}

File: src/cpp/highflyers-protocol/Parser.cpp
/*
 * highflyers-cpp-protocol
 *     Parser.cpp
 *
 *  Created on: 10 mar 2014
 *     Author: Marcin Kolny <marcin.kolny@gmail.com>
 */

#include "Parser.h"

using namespace HighFlyers::Protocol;

void Parser::append_bytes(const std::vector<byte>& data)
{
	for (auto b : data)
		append_byte(b);
}

void Parser::append_byte(byte b)
{
	if (prev_sentinel)
    {
		if (b == Sentinel || b == EndFrame)
			bytes.push_back(b);
		else
			throw new std::logic_error("Unexpected token " + b);
			
		prev_sentinel = false;
	}
	else if (b == Sentinel)
		prev_sentinel = true;
	else
		bytes.push_back(b);

	if (b == EndFrame)
		parse_frame();
	else if (MaxLength == bytes.size())
    {
		bytes.clear();
		// todo log warning or sth. No warning exception defined: throw new std::warning_exception("Too many bytes without end_frame sign. Dropping data...");
	}
}

void Parser::new_frame_received(std::shared_ptr<Frame> frame)
{
	// todo notifier!
}

void Parser::parse_frame()
{
	std::shared_ptr<Frame> frame(FrameBuilder::build_frame(bytes));
	
	new_frame_received(frame);
}
